File: R/linear_regression.R
#add_ones(X) returns the matrix (1:X), obtained by appending a column of 1's to the front of X
add_ones = function(x){
    x0 = rep(1,dim(x)[1])
    return(cbind(x0,x))
}

mlr = function(x,y){
    #Replace X with the augmented matrix (1:X) and then solve the equation
    #X^T X b = X^T y for b (the model coefficiens) with qr.solve
    x = add_ones(x)
    coeffs = qr.solve(t(x) %*% x, t(x) %*% y)
    
    #Compute details about the model fit
    train_preds = x %*% coeffs
    RSS = sum((y - train_preds)^2)
    TSS = sum((y-mean(y))^2)
    RSQ = 1 - RSS/TSS
    MSE = RSS/dim(x)[1]
    
    return(list(coeffs = coeffs, 
                rss = RSS, 
                tss = TSS, 
                rsq = RSQ, 
                mse = MSE, 
                train_preds = train_preds)
          )
}

#use model (output from mlr()) to make predictions for x_test
predict.mlr = function(model, x_test){
    x_test = add_ones(x_test)
    return(x_test %*% model$coeffs)
}

File: Python/linear_regression.py
import numpy as np
import pandas as pd

#insert_ones(X) returns the matrix (1,X), obtained  by appending a column of 1's to the front of X
def insert_ones(X):
    return np.c_[np.ones(X.shape[0]), X]

class MLR():
    """Multivariate linear regression model. Models the (linear) relationship between 
    response vector Y and predictor matrix X.
    """

    def __init__(self):
        pass
    
    def fit(self,X,Y):
        #Replace X with the augmented matrix (1,X) and then compute the QR decomposition of X^T X
        X = insert_ones(X)
        Q, R = np.linalg.qr(X.T @ X)
    
        self.coeffs = np.linalg.inv(R) @ (X@Q).T @ Y
        
        train_preds = X @ self.coeffs
        
        self.rss = ((Y-train_preds)**2).sum()
        self.tss = ((Y - Y.mean())**2).sum()
        self.rsq = 1 - self.rss/self.tss
        self.mse = self.rss/X.shape[0]
        
    def predict(self,X):
        X = insert_ones(X)
        return(X @ self.coeffs)
    
    def summary(self):
        print("Coefficients: ", self.coeffs)
        print("RSS: %f \nTSS: %f \nRsq: %f \nMSE: %f \n" % (self.rss, self.tss, self.rsq, self.mse))
